build/Dockerfile

```
@@ -1,133 +1,217 @@
 # (c) Cartesi and individual authors (see AUTHORS)
 # SPDX-License-Identifier: Apache-2.0 (see LICENSE)
 
-# syntax=docker.io/docker/dockerfile:1.4
-
-ARG RUST_VERSION=1.74.0
-ARG GO_VERSION=1.21.1
-ARG SERVER_MANAGER_VERSION=0.8.2
-ARG ROLLUPS_CONTRACTS_VERSION=1.1.0
-ARG MACHINE_EMULATOR_VERSION=0.15.2
-
-ARG BASE_PATH=/opt/cartesi
-ARG RUST_BUILD_PATH=${BASE_PATH}/src/rollups-node/offchain
-ARG GO_BASE_PATH=/go
-ARG GO_BUILD_PATH=${GO_BASE_PATH}/cartesi-rollups-node
-ARG GO_BIN_PATH=${GO_BASE_PATH}/bin
-ARG RUNTIME_DIR=/var/opt/cartesi
-ARG MACHINE_SNAPSHOT_PATH=/var/opt/cartesi/machine-snapshots
-ARG DEVNET_SHARE_PATH=/usr/share/devnet
-ARG ANVIL_IP_ADDR=0.0.0.0
-
+# syntax=docker.io/docker/dockerfile:1
+
+# This dockerfile contains multiple stages to build three final targets.
+# The file was split in a section for each final target.
+
+# Version config that should be set in the bake file.
+ARG BASE_IMAGE
+ARG RUST_VERSION
+ARG GO_VERSION
+ARG FOUNDRY_COMMIT_VERSION
+ARG ROLLUPS_CONTRACTS_VERSION
+ARG SERVER_MANAGER_VERSION
+ARG MACHINE_EMULATOR_VERSION
+ARG ROOTFS_VERSION
+ARG LINUX_VERSION
+ARG LINUX_KERNEL_VERSION
+ARG ROM_VERSION
+
+# Build directories.
+ARG MACHINE_SNAPSHOT_BUILD_PATH=/build/machine-snapshot
+ARG DEVNET_BUILD_PATH=/build/devnet
+ARG RUST_BUILD_PATH=/build/rollups-node/rust
+ARG GO_BUILD_PATH=/build/rollups-node/go
+
+# Runtime dir for the cartesi-machine snapshot.
+ARG MACHINE_SNAPSHOT_RUNTIME_PATH=/usr/share/cartesi/snapshots
+
+####################################################################################################
+# STAGE: emulator-base
 #
-# Machine snapshot generation
+# This stage creates a base-image with the Cartesi machine emulator.
+# The result is used as the base for the snapshot and the node targets.
+# We do this instead of using the cartesi/machine-emulator image to have control over the distro
+# used by the base image.
+FROM ${BASE_IMAGE} as emulator-base
+
+# Install machine-emulator
+ARG MACHINE_EMULATOR_VERSION
+RUN <<EOF
+apt-get update
+apt-get install -y curl
```

```suggestion
apt-get install -y --no-install-recommends
    ca-certificates \
    curl
```

We save 4 packages, but I won´t block for this. :)